process/block/baseProcess.go

```diff
@@ -951,39 +951,95 @@ func (bp *baseProcessor) checkHeaderBodyCorrelation(miniBlockHeaders []data.Mini
 		return process.ErrHeaderBodyMismatch
 	}
 
+	var mbHdr data.MiniBlockHeaderHandler
+	var miniBlock *block.MiniBlock
 	for i := 0; i < len(body.MiniBlocks); i++ {
-		miniBlock := body.MiniBlocks[i]
+		miniBlock = body.MiniBlocks[i]
+		mbHdr = miniBlockHeaders[i]
 		if miniBlock == nil {
 			return process.ErrNilMiniBlock
 		}
+		if mbHdr == nil {
+			return process.ErrNilMiniBlockHeader
+		}
 
 		mbHash, err := core.CalculateHash(bp.marshalizer, bp.hasher, miniBlock)
 		if err != nil {
 			return err
 		}
 
-		mbHdr, ok := mbHashesFromHdr[string(mbHash)]
-		if !ok {
-			return process.ErrHeaderBodyMismatch
+		err = checkMiniBlockWithMiniBlockHeader(mbHash, miniBlockHeaders[i], miniBlock)
+		if err != nil {
+			return err
 		}
+	}
 
-		if mbHdr.GetTxCount() != uint32(len(miniBlock.TxHashes)) {
-			return process.ErrHeaderBodyMismatch
+	return bp.checkMiniBlocksConstructionProposal(miniBlockHeaders)
+}
+
+func (bp *baseProcessor) checkMiniBlocksConstructionProposal(miniBlockHeaders []data.MiniBlockHeaderHandler) error {
+	for i := 0; i < len(miniBlockHeaders); i++ {
+		// for Supernova all miniBlocks not part of an execution result need to have construction state Proposed
+		if miniBlockHeaders[i].GetConstructionState() != int32(block.Proposed) {
+			return process.ErrWrongMiniBlockConstructionState
 		}
+		if miniBlockHeaders[i].GetProcessingType() != int32(block.Normal) {
+			return process.ErrWrongMiniBlockProcessingType
+		}
+	}
+	return nil
+}
+
+func checkMiniBlockWithMiniBlockHeader(mbHash []byte, mbHdr data.MiniBlockHeaderHandler, miniBlock *block.MiniBlock) error {
+	if !bytes.Equal(mbHash, mbHdr.GetHash()) {
+		return process.ErrHeaderBodyMismatch
+	}
+
+	if mbHdr.GetTxCount() != uint32(len(miniBlock.TxHashes)) {
+		return process.ErrHeaderBodyMismatch
+	}
+
+	if mbHdr.GetReceiverShardID() != miniBlock.ReceiverShardID {
+		return process.ErrHeaderBodyMismatch
+	}
 
-		if mbHdr.GetReceiverShardID() != miniBlock.ReceiverShardID {
-			return process.ErrHeaderBodyMismatch
+	if mbHdr.GetSenderShardID() != miniBlock.SenderShardID {
+		return process.ErrHeaderBodyMismatch
+	}
```

Comment:
It will not be very obvious what's the source of the error. Some logging might be useful. Can be left as TODO for future PRs (if really needed).

Here / below etc.